spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/RewritableDeletes.java

```diff
@@ -0,0 +1,104 @@
+import org.apache.iceberg.DeleteFile;
+import org.apache.iceberg.FileMetadata;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Iterables;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.util.DeleteFileSet;
+
+/** Wraps a set of rewritable delete files which are relativized to a given table location */
+class RewritableDeletes implements Serializable {
+  private final String tableLocation;
+  private final Map<String, DeleteFileSet> relativizedRewritableDeletes;
+
+  RewritableDeletes(String tableLocation) {
+    this.tableLocation = tableLocation;
+    this.relativizedRewritableDeletes = Maps.newHashMap();
+  }
+
+  /**
+   * @param dataFileLocation absolute data file location
+   * @param deleteFile delete file with absolute data file location
+   * @param specs map from spec ID to partition spec
+   */
+  public void addRewritableDelete(
+      String dataFileLocation, DeleteFile deleteFile, Map<Integer, PartitionSpec> specs) {
+    relativizedRewritableDeletes
+        .computeIfAbsent(relativizePath(dataFileLocation), ignored -> DeleteFileSet.create())
+        .add(relativizeDeleteFile(deleteFile, specs));
+  }
+
+  /**
+   * @return map of data to rewritable delete files with relativized paths
+   */
+  public Map<String, DeleteFileSet> relativizedDeletes() {
+    return relativizedRewritableDeletes;
+  }
+
+  /**
+   * Return an iterable of deletes for a given absolute data file location
+   *
+   * @param dataFileLocation absolute data file location
+   * @param specs map from spec ID to partition spec
+   */
+  public Iterable<DeleteFile> deletesFor(
+      String dataFileLocation, Map<Integer, PartitionSpec> specs) {
+    DeleteFileSet relativizedDeletes =
+        relativizedRewritableDeletes.get(relativizePath(dataFileLocation));
+    if (relativizedDeletes == null) {
+      return null;
+    }
+
+    return Iterables.transform(
+        relativizedDeletes,
+        deleteFile ->
+            FileMetadata.deleteFileBuilder(specs.get(deleteFile.specId()))
+                .copy(deleteFile)
+                .withPath(absolutePath(deleteFile.location()))
+                .build());
+  }
+
+  private String absolutePath(String path) {
+    return tableLocation + path;
+  }
+
+  private String relativizePath(String path) {
+    int indexOfRelativeLocation = path.indexOf(tableLocation);
+    Preconditions.checkArgument(
+        indexOfRelativeLocation != -1,
+        "Cannot relativize path: %s in relative location %s",
+        path,
+        tableLocation);
+    return path.substring(indexOfRelativeLocation + tableLocation.length());
+  }
+
+  private DeleteFile relativizeDeleteFile(
+      DeleteFile deleteFile, Map<Integer, PartitionSpec> specs) {
+    return FileMetadata.deleteFileBuilder(specs.get(deleteFile.specId()))
+        .copy(deleteFile)
```

Comment:
I think we'd have a slight perf hit by having to do this copy (particularly the copying the partition data). I'll see if there's any relevant benchmarking we can leverage to measure if it's really significant or not.

Reply:
Maybe there's a way to extend planning APIs so that we return all paths already relativized but that seems like it would be a bigger change, and it's not obvious we should do that until there's evidence that this copy is expensive.